pravega-client-examples/src/test/java/io/pravega/example/streamprocessing/WorkerProcess.java

```diff
@@ -0,0 +1,136 @@
+
+    private final WorkerProcessConfig config;
+    private final int instanceId;
+
+    private final AtomicReference<AtLeastOnceProcessorInstrumented> processor = new AtomicReference<>();
+
+    // Create the input, output, and state synchronizer streams (ignored if they already exist).
+    public void init() {
+        log.info("init: BEGIN");
+        try (StreamManager streamManager = StreamManager.create(config.clientConfig)) {
+            streamManager.createScope(config.scope);
+            final StreamConfiguration streamConfig = StreamConfiguration.builder()
+                    .scalingPolicy(ScalingPolicy.fixed(config.numSegments)).build();
+            streamManager.createStream(config.scope, config.inputStreamName, streamConfig);
+            streamManager.createStream(config.scope, config.outputStreamName, streamConfig);
+            streamManager.createStream(
+                    config.scope,
+                    config.membershipSynchronizerStreamName,
+                    StreamConfiguration.builder().scalingPolicy(ScalingPolicy.fixed(1)).build());
+        }
+        final ReaderGroupConfig readerGroupConfig = ReaderGroupConfig.builder()
+                .stream(Stream.of(config.scope, config.inputStreamName))
+                .automaticCheckpointIntervalMillis(config.checkpointPeriodMs)
+                .build();
+        try (ReaderGroupManager readerGroupManager = ReaderGroupManager.withScope(config.scope, config.clientConfig)) {
+            // Create the Reader Group (ignored if it already exists)
+            readerGroupManager.createReaderGroup(config.readerGroupName, readerGroupConfig);
+        }
+        log.info("init: END");
+    }
+
+    @Override
+    protected void run() throws Exception {
+        Serializer<TestEvent> serializer = new JSONSerializer<>(new TypeToken<TestEvent>() {}.getType());
+        try (ReaderGroupManager readerGroupManager = ReaderGroupManager.withScope(config.scope, config.clientConfig)) {
+            final ReaderGroup readerGroup = readerGroupManager.getReaderGroup(config.readerGroupName);
+            // Create client factories.
+            try (EventStreamClientFactory eventStreamClientFactory = EventStreamClientFactory.withScope(config.scope, config.clientConfig);
+                 SynchronizerClientFactory synchronizerClientFactory = SynchronizerClientFactory.withScope(config.scope, config.clientConfig);
+                 // Create a Pravega stream writer that we will send our processed output to.
+                 EventStreamWriter<TestEvent> writer = eventStreamClientFactory.createEventWriter(
+                         config.outputStreamName,
+                         serializer,
+                         EventWriterConfig.builder().build())) {
+
+                final AtLeastOnceProcessorInstrumented proc = new AtLeastOnceProcessorInstrumented(
+                        () -> ReaderGroupPruner.create(
```

Comment:
Why does this take a function instead of the value?

Reply:
It must take a supplier function for readerSupplier because the run method (https://github.com/pravega/pravega-samples/blob/f329c6b7efd17713998d085a5ac8a26f33b66a44/pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/AtLeastOnceProcessor.java#L55) needs to carefully create and close it.
As far as prunerSupplier, this could potentially be supplied directly by the caller as a value and the behavior will be correct.  Although not critical, by using a supplier, this can ensure that the pruner thread does not start until after the run() method starts. Additionally, I like the consistency in having the user provide suppliers for both readerSupplier and prunerSupplier.